### PR TITLE
Upgrade rails from 4.2.11 to 6.0

### PR DESCRIPTION
**Pull request by antima-gupta**
### Passing build links:
- [4288](https://buildkite.com/chef/chef-chef-server-master-verify/builds/4288)
## Upgrade rails from 4.2.11 to 6.0.
- Added compiled assets

### Issues Resolved

#2182 

### Check List

- [ ] New functionality includes tests
- [x] All buildkite tests pass
- [x] Full omnibus build and tests in buildkite pass
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>
- [x] PR title is a worthy inclusion in the CHANGELOG

### CODE REVIEW

File: src/oc-id/app/controllers/profiles_controller.rb [hunk 1]
```diff
@@ -21,10 +21,13 @@ def update
     # Note that if an email address is provided, don't update it right away.
     # Instead send a confirmation email and let that email link update the
     # email address.
-    #update note
-    params[:user].reject! { |k| k == :email }
+    if user_params.key?(:email) && @user.email != user_params[:email]
```
Comment (PrajaktaPurohit) [inline src/oc-id/app/controllers/profiles_controller.rb:24]: https://github.com/chef/chef-server/pull/2329 is the source for these changes. Let us not revert it.

**Comment by PrajaktaPurohit** [comment]
https://github.com/chef/chef-server/pull/2329 - needed to make some change related to fixing a vulnerability. Let us not revert these changes inadvertently.

File: src/oc-id/app/controllers/profiles_controller.rb [hunk 2]
```diff
@@ -22,9 +22,9 @@ def update
     # Instead send a confirmation email and let that email link update the
     # email address.
     #update note
-    params[:user].reject! { |k| k == :email }
+    params[:user].reject! { |k| k == "email" }
 
-    if @user.update_attributes(params[:user])
+    if @user.update_attributes(user_params)
```
Comment (PrajaktaPurohit) [inline src/oc-id/app/controllers/profiles_controller.rb:27]: Do we need this change as a part of the rails upgrade?

Comment (antima-gupta) [reply src/oc-id/app/controllers/profiles_controller.rb:27]: yes, we need to do this for rails upgrade.

File: oc-chef-pedant/spec/api/oc_id_spec.rb [hunk 3]
```diff
@@ -51,7 +51,7 @@
 
   def signin(with_callback = true)
     response = get("#{platform.server}/id/signin", oc_id_user, headers: {"Accept" => "text/html"})
-    cookie = response.headers[:set_cookie][1].split(";").first
+    cookie = response.headers[:set_cookie][0].split(";").first
```
Comment (PrajaktaPurohit) [inline oc-chef-pedant/spec/api/oc_id_spec.rb:54]: a comment about the change with respect to the rails version would be useful here.

File: src/oc-id/app/controllers/profiles_controller.rb [hunk 4]
```diff
@@ -22,9 +22,9 @@ def update
     # Instead send a confirmation email and let that email link update the
     # email address.
     #update note
-    params[:user].reject! { |k| k == :email }
+    params[:user].reject! { |k| k == "email" }
```
Comment (PrajaktaPurohit) [inline src/oc-id/app/controllers/profiles_controller.rb:25]: this probably needs a rebase.

File: src/oc-id/app/controllers/profiles_controller.rb [hunk 5]
```diff
@@ -115,4 +115,8 @@ def valid_signature?
       params[:email],
     ).valid_for?(params[:signature])
   end
+
+  def user_params
+    params.require(:user).permit(:username,:first_name,:last_name,:middle_name,:email,:public_key,:private_key,:display_name,:password)
```
Comment (PrajaktaPurohit) [inline src/oc-id/app/controllers/profiles_controller.rb:120]: should email be included in the permit list here?

File: src/oc-id/config/application.rb [hunk 6]
```diff
@@ -33,7 +33,7 @@ class Application < Rails::Application
       Doorkeeper::AuthorizationsController.layout 'application'
       Doorkeeper::AuthorizedApplicationsController.layout 'application'
     end
-
+    config.public_file_server.enabled = true
```
Comment (PrajaktaPurohit) [inline src/oc-id/config/application.rb:36]: is this required for caching? nginx does caching for server,  could you elaborate a bit more about why this needs to be set?

File: .license_scout.yml [hunk 7]
```diff
@@ -138,6 +138,9 @@ fallbacks:
     - name: pry-remote
       license_id: MIT
       license_content: https://raw.githubusercontent.com/emerose/pbkdf2-ruby/master/LICENSE.TXT
+    - name: sqlite3-ruby
```
Comment (PrajaktaPurohit) [inline .license_scout.yml:141]: where are these deps being used?

Comment (PrajaktaPurohit) [reply .license_scout.yml:141]: these are a part of rails 6! that's that then!

**Comment by PrajaktaPurohit** [comment]
Left some questions inline. Looks great overall. Could you please link to the latest adhoc and umbrella builds?

File: src/oc-id/app/controllers/profiles_controller.rb (hunk 4 above)
Comment (antima-gupta) [reply src/oc-id/app/controllers/profiles_controller.rb:27]: yes

File: src/oc-id/config/application.rb (hunk 6 above)
Comment (antima-gupta) [reply src/oc-id/config/application.rb]: I have remove this.

**Comment by PrajaktaPurohit** [comment]
Adhoc: https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2838#7afca397-801b-4b8c-8f19-9e376d2aa406

**Pull request merged**